msdk-io-mgf/src/main/java/io/github/msdk/io/mgf/MgfFileImportMethod.java

```diff
@@ -0,0 +1,162 @@
+  private long processedSpectra = 0;
+  private Hashtable<String, Pattern> patterns;
+  PriorityQueue<Pair<Double, Float>> mzIntensivePQ;
+
+  @Nullable
+  @Override
+  public Collection<MgfMsSpectrum> execute() throws MSDKException {
+    logger.info("Started MGF import from {} file", target);
+
+    try (final BufferedReader reader = new BufferedReader(new FileReader(target))) {
+      String line;
+      while ((line = reader.readLine()) != null) {
+        switch (line) {
+          case "BEGIN IONS":
+            spectra.add(processSpectrum(reader));
+            processedSpectra++;
+            break;
+        }
+      }
+      reader.close();
+    } catch (IOException e) {
+      throw new MSDKException(e);
+    }
+
+    return spectra;
+  }
+
+  private MgfMsSpectrum processSpectrum(BufferedReader reader) throws IOException {
+    String line;
+    String title = "";
+    int precursorCharge = 0;
+    double precursorMass = 0;
+    String groupped;
+
+    while (true) {
+      line = reader.readLine();
+      if (line == null || line.equals("END IONS"))
+        break;
+
+      if (line.contains("PEPMASS")) {
+        groupped = patterns.get("PEPMASS").matcher(line).group();
+        precursorMass = Double.parseDouble(groupped);
+      } else if (line.contains("TITLE")) {
+        groupped = patterns.get("TITLE").matcher(line).group();
+        title = groupped;
+      } else if (line.contains("CHARGE")) {
+        groupped =  patterns.get("CHARGE").matcher(line).group();
+        String trimmed = groupped.substring(0, groupped.length() - 1);
+        precursorCharge = Integer.parseInt(trimmed);
+        if (groupped.charAt(groupped.length() - 1) == '-') {
+          precursorCharge *= -1;
+        }
+      } else {
+        String[] floats = line.split(" ");
+        double mzValue = Double.parseDouble(floats[0]);
+        float intensiveValue = Float.parseFloat(floats[1]);
+        mzIntensivePQ.add(new Pair<>(mzValue, intensiveValue));
+      }
+    }
+
+
+    /* AbstractMsSpectrum requires sorted sequence of mz, PQ sorts values */
+    int index = 0;
+    double mz[] = new double[16];
+    float intensive[] = new float[16];
+    while (!mzIntensivePQ.isEmpty()){
+      Pair<Double, Float> pair = mzIntensivePQ.remove();
+      double mzValue = pair.getKey();
+      float intensiveValue = pair.getValue();
+      mz = ArrayUtil.addToArray(mz, mzValue, index);
+      intensive = ArrayUtil.addToArray(intensive, intensiveValue, index);
+      index++;
+    }
+
+    /*
+     Do not like this code
+     May be implement a Builder pattern for this?
+    */
+    MgfMsSpectrum spectrum = new MgfMsSpectrum(mz, intensive, index - 1, title, precursorCharge,
+        precursorMass, MsSpectrumType.CENTROIDED);
+
+    return spectrum;
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public Float getFinishedPercentage() {
+    int totalSpectra = spectra.size();
+    return totalSpectra == 0 ? null : (float) (processedSpectra / totalSpectra);
```

Comment:
This will not work, because the processedSpectra will always be equal to totalSpectra. I think in this case it is difficult to estimate the progress (the only way would be to track number of bytes read and divide by the file size), so you can just return null.